frontend/src/components/NewRunParameters.tsx

```diff
@@ -39,23 +61,150 @@ class NewRunParameters extends React.Component<NewRunParametersProps> {
         <div>{titleMessage}</div>
         {!!initialParams.length && (
           <div>
-            {initialParams.map((param, i) => (
-              <TextField
-                id={`newRunPipelineParam${i}`}
-                key={i}
-                variant='outlined'
-                label={param.name}
-                value={param.value || ''}
-                onChange={ev => handleParamChange(i, ev.target.value || '')}
-                style={{ maxWidth: 600 }}
-                className={commonCss.textField}
-              />
-            ))}
+            {initialParams.map((param, i) => {
+              return (
+                <EnhancedTextField
+                  key={i}
+                  param={param}
+                  index={i}
+                  handleParamChange={handleParamChange}
+                />
+              );
+            })}
           </div>
         )}
       </div>
     );
   }
 }
 
+interface EnhancedTextFieldProps {
+  param: ApiParameter;
+  index: number;
+  handleParamChange: (index: number, value: string) => void;
+}
+
+interface EnhancedTextFieldState {
+  isEditorOpen: boolean;
+  isInJsonForm: boolean;
+  isJsonField: boolean;
+}
+
+class EnhancedTextField extends React.Component<EnhancedTextFieldProps, EnhancedTextFieldState> {
+
+  public static getDerivedStateFromProps(nextProps: EnhancedTextFieldProps, prevState: EnhancedTextFieldState): EnhancedTextFieldState {
```

Comment:
Refer to https://reactjs.org/docs/react-component.html#static-getderivedstatefromprops

I don't see a strong reason to use `getDerivedStateFromProps`, you can just put these in render function. That can probably make it easier to understand.

Reply:
I assumed that once an editor option has been added to the Textfield then that editor option would always remain regardless of if the current input was a json string or not. My reasoning for using `getDerivedStateFromProps` is that we need to update the state if this is the first time a json string has been identified, but doing so in the render would cause the render to be impure. Do you think there is a better lifecycle method for this or do you still want me to put it in the render function?

Reply:
The behavior is well thought over. I agree with it too.

I overlooked getDerivedState using previous state. Thanks for the explanation, this should be a suitable use case.